docs/WritingADeviceDriver.md

```diff
@@ -0,0 +1,220 @@
+
+To be able to handle interrupts, you must have a [software capability](SoftwareCapabilities.md) that [authorises access to the interrupt](Interrupts.md).
+For the ethernet device that we've been using as an example, you would typically request one with this macro invocation:
+
+```c
+DECLARE_AND_DEFINE_INTERRUPT_CAPABILITY(ethernetInterruptCapability, Ethernet, true, true);
+```
+
+If you wish to share this between multiple compilation units, you can use the separate `DECLARE_` and `DEFINE_` forms (see [`interrupt.h`](../sdk/include/interrupt.h)) but the combined form is normally most convenient.
+This macro takes four arguments:
+
+ 1. The name that we're going to use to refer to this capability.
+    The name `ethernetInterruptCapability` is arbitrary, you can use whatever makes sense to you.
+ 2. The name of the interrupt, from the board description file (`Ethernet`, in this case).
+ 3. Whether this capability authorises waiting for this interrupt (this will almost always be `true`).
+ 4. Whether this capability authorises acknowledging the interrupt so that it can fire again.
+    This will almost always be true in device drivers but should generally be true for only one compartment (for each interrupt), whereas multiple compartments may wish to observe interrupts for monitoring.
+
+As with the MMIO capabilities, compartments 
+
+```json
+        {
+          "contents": "10000101",
+          "kind": "SealedObject",
+          "sealing_type": {
+            "compartment": "sched",
+            "key": "InterruptKey",
+            "provided_by": "build/cheriot/cheriot/release/example-firmware.scheduler.compartment",
+            "symbol": "__export.sealing_type.sched.InterruptKey"
+          }
+```
+
+The sealing type tells you that this is an interrupt capability (it's sealed with the `InterruptKey` type, provided by the scheduler).
+The contents lets you audit what this authorises.
+The first two bytes are a 16-bit (little-endian on all currently supported targets) integer containing the interrupt number, so 1000 means 16 (our Ethernet interrupt number).
+The next two bytes are boolean values reflecting the last two arguments to the macro, so this authorises both waiting and clearing the macro.
+Again, this can form part of your firmware auditing.
+
+### Waiting on an interrupt
+
+Now that you're authorised to handle interrupts, you will need something that you can wait on.
+Each interrupt is mapped to a futex word, which can be used with scheduler waiting primitives.
+You can get the word associated with an interrupt by passing the authorising capability to the `interrupt_futex_get` function exported by the scheduler:
+
+```c
+const uint32_t *ethernetFutex = ethernetFutex = interrupt_futex_get(STATIC_SEALED_VALUE(ethernetInterruptCapability));
+```
+
+The `ethernetFutex` pointer is now a read-only capability (attempting to store through it will trap) that contains a number that is incremented every time that an interrupt fires.
```

Comment:
```suggestion
The `ethernetFutex` pointer is now a read-only capability (attempting to store through it will trap) that contains a number that is incremented every time the ethernet interrupt fires.
```